### PR TITLE
[API 2.0] add functional pool1d API

### PR DESCRIPTION
**Pull request by LDOUBLEV**
### PR types

New features

### PR changes

APIs

### Describe

add  functional APIs:
- paddle.nn.functional.avg_pool1d
- paddle.nn.functional.max_pool1d
- paddle.nn.functional.adaptive_avg_pool1d
- paddle.nn.functional.adaptive_max_pool1d
- paddle.nn.AvgPool1d
- paddle.nn.MaxPool1d
- paddle.nn.AdaptiveAvgPool1d
- paddle.nn.AdaptiveMaxPool1d

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_pool1d_op.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+# limitations under the License.
+
+import numpy as np
+
+import paddle.fluid as fluid
+from paddle.nn.functional.pooling import avg_pool1d, max_pool1d, adaptive_max_pool1d, adaptive_avg_pool1d
+
+
+def adaptive_start_index(index, input_size, output_size):
+    return int(np.floor(index * input_size / output_size))
```
Comment (littletomatodonkey) [inline python/paddle/fluid/tests/unittests/test_pool1d_op.py:28]: 直接int就可以？

File: python/paddle/fluid/tests/unittests/test_pool1d_op.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+
+def adaptive_end_index(index, input_size, output_size):
+    return int(np.ceil((index + 1) * input_size / output_size))
+
+
+def max_pool1D_forward_naive(x,
```
Comment (littletomatodonkey) [inline python/paddle/fluid/tests/unittests/test_pool1d_op.py:35]: pool1D -> pool1d ?

File: python/paddle/fluid/tests/unittests/test_pool1d_op.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+        elif pool_type == 'max':
+            out[:, :, i] = np.max(x_masked, axis=(2))
+
+    return out
+
+
+def test_np_pd_pool1d():
+    data = np.random.random((2, 4, 32)).astype('float32')
+    with fluid.dygraph.guard():
+        # datapd = fluid.layers.assign(data)
```
Comment (littletomatodonkey) [inline python/paddle/fluid/tests/unittests/test_pool1d_op.py]: comment删掉？

File: python/paddle/fluid/tests/unittests/test_pool1d_op.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+        datapd = fluid.dygraph.to_variable(data)
+
+        res_pd = max_pool1d(
+            datapd, kernel_size=3, stride=2, padding=0, ceil_mode=False)
+
+        res_np = max_pool1D_forward_naive(
+            data, ksize=[3], paddings=[0], strides=[2], ceil_mode=False)
+
+    np.testing.assert_allclose(res_np, res_pd.numpy())
+    print("=> unittest avgpool1d success!")
```
Comment (littletomatodonkey) [inline python/paddle/fluid/tests/unittests/test_pool1d_op.py]: print删掉

File: python/paddle/fluid/tests/unittests/test_pool1d_op.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+    with fluid.dygraph.guard():
+        datapd = fluid.dygraph.to_variable(data)
+
+        res_pd = adaptive_avg_pool1d(datapd, 6)
+
+        res_np = avg_pool1D_forward_naive(
+            data, ksize=[6], adaptive=True, paddings=[0], strides=[0])
+
+    np.testing.assert_allclose(res_np, res_pd.numpy())
+    print("=> unittest adaptive_avg_pool1d success!")
```
Comment (littletomatodonkey) [inline python/paddle/fluid/tests/unittests/test_pool1d_op.py]: 同上

File: python/paddle/nn/functional/pooling.py [hunk 6]
```diff
@@ -17,5 +17,479 @@
...
+        ShapeError: If the output's shape calculated is not greater than 0.
+
+
+    Examples:
+
+        .. code-block:: python
+
+          import paddle.fluid as fluid
+
+          data = fluid.data(name='data', shape=[None, 3, 32], dtype='float32')
```
Comment (littletomatodonkey) [inline python/paddle/nn/functional/pooling.py]: paddle.data?2.0好像主要用paddle.了

File: python/paddle/nn/functional/pooling.py [hunk 7]
```diff
@@ -17,5 +17,479 @@
...
+
+          # max pool2d
+          pool2d = fluid.layers.functional.avg_pool1d(input=data, kernel_size=2, stride=2, padding=0)
```
Comment (littletomatodonkey) [inline python/paddle/nn/functional/pooling.py]: paddle.nn.functional

File: python/paddle/fluid/tests/unittests/test_pool1d_op.py (hunk 1 above)
Comment (LDOUBLEV) [reply python/paddle/fluid/tests/unittests/test_pool1d_op.py:28]: 为了和ceil mode区分

File: python/paddle/fluid/tests/unittests/test_pool1d_op.py (hunk 3 above)
Comment (LDOUBLEV) [reply python/paddle/fluid/tests/unittests/test_pool1d_op.py]: Done

File: python/paddle/nn/functional/pooling.py (hunk 6 above)
Comment (LDOUBLEV) [reply python/paddle/nn/functional/pooling.py]: 应该还是fluid.data，可以再确认下

File: python/paddle/nn/functional/pooling.py (hunk 7 above)
Comment (LDOUBLEV) [reply python/paddle/nn/functional/pooling.py]: Done！THX

File: python/paddle/nn/functional/pooling.py [hunk 8]
```diff
@@ -13,9 +13,519 @@
...
+
+    return padding
+
+
+def avg_pool1d(x,
+               kernel_size,
+               stride=None,
+               padding=0,
+               count_include_pad=True,
+               ceil_mode=False,
```
Comment (WuHaobo) [inline python/paddle/nn/functional/pooling.py:63]: data_format='NCL'

File: python/paddle/nn/functional/pooling.py [hunk 9]
```diff
@@ -13,9 +13,519 @@
...
+            it could be the following forms: `[pad_left, pad_right]`. If padding is non-zero,
+            then the input is implicitly zero-padded on both sides for padding number of points.
+        ceil_mode (bool): ${ceil_mode_comment}Whether to use the ceil function to calculate output height and width.
+            If it is set to False, the floor function will be used. Default False
+        name(str, optional): For detailed information, please refer
+                             to :ref:`api_guide_Name`. Usually name is no need to set and
+                             None by default.
+        count_include_pad (bool): Whether to exclude padding points in average pooling
+                          mode, default is `true`.
+
```
Comment (WuHaobo) [inline python/paddle/nn/functional/pooling.py:102]: Args 顺序，name最后一个

File: python/paddle/nn/functional/pooling.py [hunk 10]
```diff
@@ -13,9 +13,519 @@
...
+    return squeeze(pool_out, [2])
+
+
+def max_pool1d(x,
+               kernel_size,
+               stride=None,
+               padding=0,
+               ceil_mode=False,
+               return_indices=False,
```
Comment (WuHaobo) [inline python/paddle/nn/functional/pooling.py:199]: return_indices=False, ceil_mode=False, data_format='NCL'

File: python/paddle/nn/layer/common.py [hunk 11]
```diff
@@ -342,3 +342,321 @@ def forward(self, input):
...
+          # pool_out shape: [1, 3, 16] 
+
+    """
+
+    def __init__(self,
+                 kernel_size,
+                 stride=None,
+                 padding=0,
+                 ceil_mode=False,
+                 count_include_pad=True):
```
Comment (WuHaobo) [inline python/paddle/nn/layer/common.py]: dataformat name

**Review by WuHaobo — APPROVED** [review]
LGTM

File: python/paddle/nn/layer/common.py [hunk 12]
```diff
@@ -28,6 +28,10 @@
     'Linear',
     'UpSample',
     'Pad2D',
+    'AvgPool1d',
```
Comment (shippingwang) [inline python/paddle/nn/layer/common.py:31]: move pool layer to layer/pooling.py ?

File: python/paddle/nn/functional/pooling.py [hunk 13]
```diff
@@ -13,9 +13,516 @@
...
+
+    ..  math::
+
+       Output(N_i, C_i, l) &= mean(Input[N_i, C_i, stride \times l:stride \times l+k])
+
+
+    Args:
+        x (Tensor): The input tensor of pooling operator which is a 3-D tensor with
+                          shape [N, C, L]. The format of input tensor is `"NCL"` or
+                          `"NHL"`, where `N` is batch size, `C` is the number of channels,
```
Comment (shippingwang) [inline python/paddle/nn/functional/pooling.py:85]: NHL -> NLC

File: python/paddle/nn/functional/pooling.py [hunk 14]
```diff
@@ -13,9 +13,516 @@
...
+          import paddle.nn.functional as F
+          paddle.disable_static()
+
+          data = paddle.to_tensor(np.random.uniform(-1, 1, [1, 3, 32]).astype(np.float32))
+          pool_out = F.avg_pool1d(data, kernel_size=2, stride=2, padding=0)
+          # pool_out shape: [1, 3, 16]
+
+    """
+    """NCL to NCHW"""
+    data_format = "NCHW"
```
Comment (shippingwang) [inline python/paddle/nn/functional/pooling.py:128]: add NHWC?

File: python/paddle/nn/functional/pooling.py [hunk 15]
```diff
@@ -13,9 +13,516 @@
...
+    padding = update_padding1d(padding, "avg")
+
+    if in_dygraph_mode():
+        output = core.ops.pool2d(
+            x, 'pooling_type', 'avg', 'ksize', kernel_size, 'global_pooling',
+            False, 'strides', stride, 'paddings', padding, 'padding_algorithm',
+            padding_algorithm, 'use_cudnn', not count_include_pad, 'ceil_mode',
+            ceil_mode, 'use_mkldnn', False, 'exclusive', True, 'data_format',
+            data_format)
+        return squeeze(output, [2])
```
Comment (shippingwang) [inline python/paddle/nn/functional/pooling.py:167]: call core.ops.squeeze?

File: python/paddle/nn/functional/pooling.py [hunk 16]
```diff
@@ -13,9 +13,516 @@
...
+    For average pool1d:
+
+    ..  math::
+
+       Output(N_i, C_i, l) &=  max(Input[N_i, C_i, stride \times l:stride \times l+k])}
+
+    Args:
+        x (Tensor): The input tensor of pooling operator which is a 3-D tensor with
+                          shape [N, C, L]. The format of input tensor is `"NCL"` or
+                          `"NHL"`, where `N` is batch size, `C` is the number of channels,
```
Comment (shippingwang) [inline python/paddle/nn/functional/pooling.py:220]: NLC

File: python/paddle/nn/functional/pooling.py [hunk 17]
```diff
@@ -13,9 +13,516 @@
...
+
+__all__ = [
+    'pool2d', 'pool3d', 'adaptive_pool2d', 'adaptive_pool3d', 'avg_pool1d',
+    'max_pool1d', 'adaptive_avg_pool1d', 'adaptive_max_pool1d'
+]
+
+
+def check_input(x, dimension):
+    if len(x.shape) != dimension:
+        raise ValueError("Excepted Input X is 3-D tensor, but received {}-D {}".
```
Comment (shippingwang) [inline python/paddle/nn/functional/pooling.py:35]: dimension - D

File: python/paddle/nn/functional/pooling.py [hunk 18]
```diff
@@ -13,9 +13,516 @@
...
+          paddle.disable_static()
+
+          data = paddle.to_tensor(np.random.uniform(-1, 1, [1, 3, 32]).astype(np.float32))
+          pool_out = F.avg_pool1d(data, kernel_size=2, stride=2, padding=0)
+          # pool_out shape: [1, 3, 16]
+
+    """
+    """NCL to NCHW"""
+    data_format = "NCHW"
+    check_variable_and_dtype(x, 'input', ['float32', 'float64'], 'avg_pool1d')
```
Comment (shippingwang) [inline python/paddle/nn/functional/pooling.py:129]: if not in_dygraph_mode

File: python/paddle/nn/layer/common.py [hunk 19]
```diff
@@ -348,6 +352,310 @@ def forward(self, input):
...
+        output_size (int|list|tuple): The pool kernel size. If pool kernel size is a tuple or list,
+            it must contain one int.
+        return_indices (bool): If true, the index of max pooling point will be returned along
+            with outputs. It cannot be set in average pooling type. Default False.
+        name(str, optional): For detailed information, please refer
+                             to :ref:`api_guide_Name`. Usually name is no need to set and
+                             None by default.
+    Returns:
+        None.
+
```
Comment (shippingwang) [inline python/paddle/nn/layer/common.py:613]: shape?

File: python/paddle/nn/layer/common.py [hunk 20]
```diff
@@ -348,6 +352,310 @@ def forward(self, input):
...
+          import paddle.nn as nn
+          paddle.disable_static()
+
+          data = paddle.to_tensor(np.random.uniform(-1, 1, [1, 3, 32]).astype(np.float32))
+          AdaptiveMaxPool1d = nn.AdaptiveMaxPool1d(output_size=16)
+          pool_out = AdaptiveMaxPool1d(data)
+          # pool_out shape: [1, 3, 16]
+
+          # for return_indices = true
+          AdaptiveMaxPool1d = nn.AdaptiveMaxPool1d(output_size=16, return_indices=True)
```
Comment (shippingwang) [inline python/paddle/nn/layer/common.py:642]: use lower name name function

File: python/paddle/nn/layer/common.py [hunk 21]
```diff
@@ -348,6 +352,310 @@ def forward(self, input):
...
+
+    ..  math::
+
+       lstart &= floor(i * L_{in} / L_{out})
+
+       lend &= ceil((i + 1) * L_{in} / L_{out})
+
+       Output(i) &= max(Input[lstart:lend])}
+
+    Args:
```
Comment (shippingwang) [inline python/paddle/nn/layer/common.py:603]: Parameters

File: python/paddle/nn/functional/pooling.py [hunk 22]
```diff
@@ -13,9 +13,516 @@
...
+            output_size (int|list|tuple): The pool kernel size. If pool kernel size is a tuple or list,
+                it must contain one int.
+            return_indices (bool): If true, the index of max pooling point will be returned along
+                with outputs. It cannot be set in average pooling type. Default False.
+            name(str, optional): For detailed information, please refer
+                                 to :ref:`api_guide_Name`. Usually name is no need to set and
+                                 None by default.
+
+        Returns:
+            Tensor: The output tensor of adaptive pooling result. The data type is same
```
Comment (shippingwang) [inline python/paddle/nn/functional/pooling.py:458]: There are two conditions

**Comment by LDOUBLEV** [comment]
为了防止PR合入冲突，将该PR 的修改 合并到 https://github.com/PaddlePaddle/Paddle/pull/26331 中，此PR可关闭

File: python/paddle/nn/layer/common.py (hunk 12 above)
Comment (LDOUBLEV) [reply python/paddle/nn/layer/common.py:31]: 目前是都在common.py里的

File: python/paddle/nn/functional/pooling.py (hunk 13 above)
Comment (LDOUBLEV) [reply python/paddle/nn/functional/pooling.py:85]: dataformat 低优

File: python/paddle/nn/functional/pooling.py (hunk 18 above)
Comment (LDOUBLEV) [reply python/paddle/nn/functional/pooling.py:129]: 适用静态图模式

File: python/paddle/nn/layer/common.py (hunk 19 above)
Comment (LDOUBLEV) [reply python/paddle/nn/layer/common.py:613]: 返回的是一个对象

**Pull request closed without merging**